File: src/main/scala/challenge/c2020_12/c2020_12_19.scala
package challenge.c2020_12

import org.scalatest.matchers.should.Matchers
import org.scalatest.wordspec.AnyWordSpec

/**
 * [[https://leetcode.com/explore/challenge/card/december-leetcoding-challenge/571/week-3-december-15th-december-21st/3571/]]
 */
//noinspection DuplicatedCode
class c2020_12_19 extends AnyWordSpec with Matchers {

  /**
   * === Cherry Pickup II ===
   *
   * Given a `rows x cols` matrix `grid` representing a field of cherries.
   * Each cell in `grid` represents the number of cherries that you can collect.
   *
   * You have two robots that can collect cherries for you, Robot #1 is located at the top-left corner `(0,0)`,
   * and Robot #2 is located at the top-right corner `(0, cols-1)` of the grid.
   *
   * Return ''the maximum number of cherries collection using both robots'' by following the rules below:
   *  - From a cell `(i,j)`, robots can move to cell `(i+1, j-1)`, `(i+1, j)` or `(i+1, j+1)`.
   *  - When any robot is passing through a cell, it picks up all cherries, and the cell becomes an empty cell `(0)`.
   *  - When both robots stay on the same cell, only one of them takes the cherries.
   *  - Both robots cannot move outside of the `grid` at any moment.
   *  - Both robots should reach the bottom row in the `grid`.
   *
   * '''Constraints:'''
   *  - `2 <= grid.length, grid[i].length <= 70`
   *  - `0 <= grid[i][j] <= 100 `
   */
  object Solution {
    def cherryPickup(G: Array[Array[Int]]): Int = {
      val rows = G.length
      val cols = G(0).length
      val dp = Array.fill(rows + 1, cols + 2, cols + 2)(0)

      for (i <- rows - 1 to 0 by -1; j <- 1 to cols; k <- 1 to cols if j != k) {
        val nMax =
          Seq(
            (j - 1, k - 1), (j - 1, k), (j - 1, k + 1),
            (j, k - 1), (j, k), (j, k + 1),
            (j + 1, k - 1), (j + 1, k), (j + 1, k + 1)
          ).map { case (j, k) => dp(i + 1)(j)(k) }
            .max
        dp(i)(j)(k) = G(i)(j - 1) + G(i)(k - 1) + nMax
      }

      dp(0)(1)(cols)
    }
  }

  import Solution.cherryPickup

  "Example 1: (grid = [[3,1,1],[2,5,1],[1,5,5],[2,1,1]]) -> 24" in {
    val grid = Array(
      Array(3, 1, 1),
      Array(2, 5, 1),
      Array(1, 5, 5),
      Array(2, 1, 1))
    cherryPickup(grid) shouldBe 24
    //Explanation:
    // Cherries taken by Robot #1, (3 + 2 + 5 + 2) = 12.
    // Cherries taken by Robot #2, (1 + 5 + 5 + 1) = 12.
    // Total of cherries: 12 + 12 = 24.
  }
  "Example 2: (grid = [[1,0,0,0,0,0,1],[2,0,0,0,0,3,0],[2,0,9,0,0,0,0],[0,3,0,5,4,0,0],[1,0,2,3,0,0,6]]) -> 28" in {
    val grid = Array(
      Array(1, 0, 0, 0, 0, 0, 1),
      Array(2, 0, 0, 0, 0, 3, 0),
      Array(2, 0, 9, 0, 0, 0, 0),
      Array(0, 3, 0, 5, 4, 0, 0),
      Array(1, 0, 2, 3, 0, 0, 6))
    cherryPickup(grid) shouldBe 28
    //Explanation: Path of robot #1 and #2 are described in color green and blue respectively.
    // Cherries taken by Robot #1, (1 + 0 + 9 + 5 + 2) = 17.
    // Cherries taken by Robot #2, (1 + 3 + 0 + 4 + 3) = 11.
    // Total of cherries: 17 + 11 = 28.
  }
  "Example 3: (grid = [[1,0,0,3],[0,0,0,3],[0,0,3,3],[9,0,3,3]]) -> 22" in {
    val grid = Array(
      Array(1, 0, 0, 3),
      Array(0, 0, 0, 3),
      Array(0, 0, 3, 3),
      Array(9, 0, 3, 3))
    cherryPickup(grid) shouldBe 22
  }
  "Example 4: (grid = [[1,1],[1,1]]) -> 4" in {
    val grid = Array(
      Array(1, 1),
      Array(1, 1))
    cherryPickup(grid) shouldBe 4
  }

  "(grid = [[1,5],[1,5]]) -> 12" in {
    val grid = Array(
      Array(1, 5),
      Array(1, 5))
    cherryPickup(grid) shouldBe 12
  }
  "(grid = [[5,1],[5,1]]) -> 12" in {
    val grid = Array(
      Array(5, 1),
      Array(5, 1))
    cherryPickup(grid) shouldBe 12
  }
  "(grid = [[1,1,1],[1,1,1],[1,1,1]]) -> 6" in {
    val grid = Array(
      Array(1, 1, 1),
      Array(1, 1, 1),
      Array(1, 1, 1))
    cherryPickup(grid) shouldBe 6
  }
  "(grid = [[1,2,3],[4,5,6],[7,8,9]]) -> 32" in {
    val grid = Array(
      Array(1, 2, 3),
      Array(4, 5, 6),
      Array(7, 8, 9))
    cherryPickup(grid) shouldBe 32
  }

  "(grid = [70 x 70 filled with 0]) -> 0" in {
    val grid = Array.fill(70, 70)(0)
    cherryPickup(grid) shouldBe 0
  }
  "(grid = [70 x 70 filled with 1]) -> 140" in {
    val grid = Array.fill(70, 70)(1)
    cherryPickup(grid) shouldBe 140
  }

}

File: ts/test/c2020_12/c2020_12_19.ts
import {expect} from 'chai';

// bottom up dp
// noinspection JSUnusedLocalSymbols
function cherryPickupBottomUpDp(grid: number[][]): number {
    const n = grid.length;
    const m = grid[0].length;

    const dp: number[][][] = [...Array(n)].map(_ => [...Array(m)].map(_ => Array(m).fill(0)));

    for (let i = n - 1; i >= 0; i--) {
        for (let r1 = 0; r1 < m - 1; r1++) {
            for (let r2 = r1 + 1; r2 < m; r2++) {
                let max_next = 0;
                if (i < n - 1) {
                    for (let nr1 = r1 - 1; nr1 <= r1 + 1; nr1++) {
                        for (let nr2 = r2 - 1; nr2 <= r2 + 1; nr2++) {
                            if (nr1 >= 0 && nr1 < nr2 && nr2 < m) {
                                max_next = Math.max(max_next, dp[i + 1][nr1][nr2]);
                            }
                        }
                    }
                }
                dp[i][r1][r2] = grid[i][r1] + grid[i][r2] + max_next;
            }
        }
    }

    return dp[0][0][m - 1];
}

// top down dp / memoization
// noinspection JSUnusedLocalSymbols
function cherryPickup(grid: number[][]): number {
    const n = grid.length;
    const m = grid[0].length;

    const dp: number[][][] = [...Array(n)].map(_ => [...Array(m)].map(_ => Array(m).fill(-1)));

    const rec = (i: number, r1: number, r2: number): number => {
        if (i == n) return 0;
        else if (dp[i][r1][r2] != -1) return dp[i][r1][r2];
        else {
            let max_next = 0;
            for (let nr1 = r1 - 1; nr1 <= r1 + 1; nr1++) {
                for (let nr2 = r2 - 1; nr2 <= r2 + 1; nr2++) {
                    if (nr1 >= 0 && nr1 < nr2 && nr2 < m) {
                        max_next = Math.max(max_next, rec(i + 1, nr1, nr2));
                    }
                }
            }

            dp[i][r1][r2] = grid[i][r1] + grid[i][r2] + max_next;
            return dp[i][r1][r2];
        }
    };

    return rec(0, 0, m - 1);
}

// noinspection JSUnusedLocalSymbols
function cherryPickupBruteForce(grid: number[][]): number {
    const n = grid.length;
    const m = grid[0].length;

    const rec = (i: number, r1: number, r2: number): number => {
        if (i == n) return 0;
        else {
            let max_next = 0;
            for (let nr1 = r1 - 1; nr1 <= r1 + 1; nr1++) {
                for (let nr2 = r2 - 1; nr2 <= r2 + 1; nr2++) {
                    if (nr1 >= 0 && nr1 < nr2 && nr2 < m) {
                        max_next = Math.max(max_next, rec(i + 1, nr1, nr2));
                    }
                }
            }

            return grid[i][r1] + grid[i][r2] + max_next;
        }
    };

    return rec(0, 0, m - 1);
}

describe('c2020_12_19 Cherry Pickup II', () => {
    it('Example 1: (grid = [[3,1,1],[2,5,1],[1,5,5],[2,1,1]]) -> 24', () => {
        const grid = [
            [3, 1, 1],
            [2, 5, 1],
            [1, 5, 5],
            [2, 1, 1]];
        expect(cherryPickup(grid)).to.eql(24);
        //Explanation:
        // Cherries taken by Robot #1, (3 + 2 + 5 + 2) = 12.
        // Cherries taken by Robot #2, (1 + 5 + 5 + 1) = 12.
        // Total of cherries: 12 + 12 = 24.
    });
    it('Example 2: (grid = [[1,0,0,0,0,0,1],[2,0,0,0,0,3,0],[2,0,9,0,0,0,0],[0,3,0,5,4,0,0],[1,0,2,3,0,0,6]]) -> 28', () => {
        const grid = [
            [1, 0, 0, 0, 0, 0, 1],
            [2, 0, 0, 0, 0, 3, 0],
            [2, 0, 9, 0, 0, 0, 0],
            [0, 3, 0, 5, 4, 0, 0],
            [1, 0, 2, 3, 0, 0, 6]];
        expect(cherryPickup(grid)).to.eql(28);
        //Explanation:
        // Cherries taken by Robot #1, (1 + 0 + 9 + 5 + 2) = 17.
        // Cherries taken by Robot #2, (1 + 3 + 0 + 4 + 3) = 11.
        // Total of cherries: 17 + 11 = 28.
    });
    it('Example 3: (grid = [[1,0,0,3],[0,0,0,3],[0,0,3,3],[9,0,3,3]]) -> 22', () => {
        const grid = [
            [1, 0, 0, 3],
            [0, 0, 0, 3],
            [0, 0, 3, 3],
            [9, 0, 3, 3]];
        expect(cherryPickup(grid)).to.eql(22);
    });
    it('Example 4: (grid = [[1,1],[1,1]]) -> 4"', () => {
        const grid = [
            [1, 1],
            [1, 1]];
        expect(cherryPickup(grid)).to.eql(4);
    });

    it('(grid = [[1,5],[1,5]]) -> 12"', () => {
        const grid = [
            [1, 5],
            [1, 5]];
        expect(cherryPickup(grid)).to.eql(12);
    });
    it('(grid = [[5,1],[5,1]]) -> 12"', () => {
        const grid = [
            [5, 1],
            [5, 1]];
        expect(cherryPickup(grid)).to.eql(12);
    });
    it('(grid = [[1,1,1],[1,1,1],[1,1,1]]) -> 6"', () => {
        const grid = [
            [1, 1, 1],
            [1, 1, 1],
            [1, 1, 1]];
        expect(cherryPickup(grid)).to.eql(6);
    });
    it('(grid = [[1,2,3],[4,5,6],[7,8,9]]) -> 32"', () => {
        const grid = [
            [1, 2, 3],
            [4, 5, 6],
            [7, 8, 9]];
        expect(cherryPickup(grid)).to.eql(32);
    });
    it('(grid = [70 x 70 filled with 0]) -> 0"', () => {
        const grid: number[][] = Array(70).fill(0).map(() => Array(70).fill(0));
        expect(cherryPickup(grid)).to.eql(0);
    });
    it('(grid = [70 x 70 filled with 1]) -> 140"', () => {
        const grid: number[][] = Array(70).fill(0).map(() => Array(70).fill(1));
        expect(cherryPickup(grid)).to.eql(140);
    });
});
